pdns/dnsdist-tcp.cc

```diff
@@ -201,544 +260,863 @@ void TCPClientCollection::addTCPClientThread()
+    if (state->d_state != IncomingTCPConnectionState::State::doingHandshake &&
+        state->d_state != IncomingTCPConnectionState::State::readingQuerySize &&
+        state->d_state != IncomingTCPConnectionState::State::readingQuery &&
+        state->d_state != IncomingTCPConnectionState::State::sendingResponse) {
+      vinfolog("Unexpected state %d in handleIOCallback", static_cast<int>(state->d_state));
+    }
+  }
+  catch(const std::exception& e) {
+    /* most likely an EOF because the other end closed the connection,
+       but it might also be a real IO error or something else.
+       Let's just drop the connection
+    */
+    if (state->d_state == IncomingTCPConnectionState::State::doingHandshake ||
+        state->d_state == IncomingTCPConnectionState::State::readingQuerySize ||
+        state->d_state == IncomingTCPConnectionState::State::readingQuery) {
+      ++state->d_ci.cs->tcpDiedReadingQuery;
+    }
+    else if (state->d_state == IncomingTCPConnectionState::State::sendingResponse) {
+      ++state->d_ci.cs->tcpDiedSendingResponse;
+    }
+
+    if (state->d_lastIOState == IOState::NeedWrite || state->d_readingFirstQuery) {
+      vinfolog("Got an exception while handling (%s) TCP query from %s: %s", (state->d_lastIOState == IOState::NeedRead ? "reading" : "writing"), state->d_ci.remote.toStringWithPort(), e.what());
+    }
+    else {
+      vinfolog("Closing TCP client connection with %s", state->d_ci.remote.toStringWithPort());
+    }
+    /* remove this FD from the IO multiplexer */
+    iostate = IOState::Done;
+  }
+
+  if (iostate == IOState::Done) {
+    handleNewIOState(state, iostate, fd, handleIOCallback);
+  }
+  else {
+    handleNewIOState(state, iostate, fd, handleIOCallback, iostate == IOState::NeedRead ? state->getClientReadTTD(now) : state->getClientWriteTTD(now));
+  }
+}
+
+static void handleIncomingTCPQuery(int pipefd, FDMultiplexer::funcparam_t& param)
+{
+  auto threadData = boost::any_cast<TCPClientThreadData*>(param);
+
+  ConnectionInfo* citmp{nullptr};
+
+  ssize_t got = read(pipefd, &citmp, sizeof(citmp));
+  if (got == 0) {
+    throw std::runtime_error("EOF while reading from the TCP acceptor pipe (" + std::to_string(pipefd) + ") in " + std::string(isNonBlocking(pipefd) ? "non-blocking" : "blocking") + " mod");
+  }
+  else if (got == -1) {
+    if (errno == EAGAIN || errno == EINTR) {
+      return;
+    }
+    throw std::runtime_error("Error while reading from the TCP acceptor pipe (" + std::to_string(pipefd) + ") in " + std::string(isNonBlocking(pipefd) ? "non-blocking" : "blocking") + " mde:" + strerror(errno));
+  }
+  else if (got != sizeof(citmp)) {
+    throw std::runtime_error("Partial read while reading from the TCP acceptor pipe (" + std::to_string(pipefd) + ") in " + std::string(isNonBlocking(pipefd) ? "non-blocking" : "blocking") + " mode");
+  }
+
+  g_tcpclientthreads->decrementQueuedCount();
+  auto ci = std::move(*citmp);
```

Comment:
I was wondering if we could get rid of `ci` and just construct `state` directly by moving `citmp`'s content